ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -491,6 +459,10 @@ void object_record_free(object_record *record) {
   ruby_xfree(record);
 }
 
+void object_record_mark(object_record *record) {
+  rb_gc_mark(record->obj_id);
+}
```

Comment:
I... suspect this may not be needed in practice. Looking at the implementation of `rb_obj_id`/`cached_object_id` the id is a regular Ruby integer that gets incremented by some number every time a new id is needed. This means this integer starts life as an immediate (e.g. not something allocated on the heap), and only if it gets too big to fit in an immediate, will it get promoted to a heap-allocated `Bignum`.

Thus, I'm thinking one potential optimization might be to check that the object id we get is indeed an immediate, and in that case we don't need to care about marking.

If it ever gets to bignum, we could either start doing the marking in such a situation (so we could have a flag that's "seen a bignum" to control the marking), or we could just not support that for now and stop heap profiling if we ever get to that -- I'm not entirely sure the exact number, but let's assume it's somewhere around 2^60 -- that's a lot of objects allocated and a very long time for an app to live.